### PR TITLE
WIP: #98 Major refactor of frontend

### LINKED ISSUE #98: Major refactor of frontend
Major frontend refactor to change architecture into something more maintainable.
Including watching for changes in rebuilding

### LINKED ISSUE #99: User registration
User registration is not working. Nothing happens after filling fields and clicking to create account.

### LINKED ISSUE #101: Set timestamp of sent messages to the time when server confirmed their receivment

### PR DESCRIPTION
**Pull request by michaelKurowski**
**References to related tickets:**
 - resolves #98 
 - resolves #99 
 - resolves #101 

**Description of changes:**
 - Abandoning react-scripts
   - Creating own build config using webpack
   - Running Tests without react-scripts
 - Changing frontend architecture into more scaleable solution
   - Separate core logic from view logic
   - Split components into smaller ones
 - Adding hypervisor for both backend and frontend changes
 - Fixing account creation
 - Adding thumbinals next to messages, found users, and room members
 - User can now send messages with the "Enter" key
 - Users finding input now autoclears after inviting users
 - Message typing input autoclears after sending message
 - Introduced Redux
 - Server now resends, sent message to sender

**Work to do**
 - Design new frontend architecture **100%**
 - Split application into smaller components **96%**
   - chatpage **98%**
   - homepage **100%**
   - register **50%**
   - create separated service for websockets **95%**
   - other **100%**
 - Make a mutual point where babel and webpack can keep babel configuration **0%**
 - Fix Jest tests **0%**
 - Add development watch mode for both backend and frontend **0%**
 - Adjust new Jest configuration to VSC to enable debugging **0%**

**New architecture**
 - Every websocket interaction should be done via sagas

### CODE REVIEW

**Comment by michaelKurowski** [comment]
TODO:
 - Design new frontend architecture
 - Split application into smaller components
 - Make a mutual point where babel and webpack can keep babel configuration
 - Fix Jest tests
 - Add hypervisor for both backend and frontend
 - Adjust new Jest configuration to VSC to enable debugging
 - Clean unused methods from chatpage.js

**Comment by michaelKurowski** [comment]
TODO: Move WS requests to saga

**Comment by michaelKurowski** [comment]
Last two small tasks to do, and then I can move to write tests:
 - Move WS requests to saga
   - create room
 - Remove redux-thunk

File: src/server/frontEnd/src/utils/sockets/ws-routing.js [hunk 1]
```diff
@@ -1,11 +1,55 @@
...
+	get
+}
+
+function inject(injectedDispatch) {
+	dispatch = injectedDispatch
+}
+
+function create() {
+	socket = {
+		room: io(`${HOST}${ROOM}`, {path: '/connection'}),
```
Comment (Jordan141) [inline src/server/frontEnd/src/utils/sockets/ws-routing.js]: Magic string, put in variable

File: src/server/frontEnd/src/utils/sockets/ws-routing.js [hunk 2]
```diff
@@ -1,11 +1,55 @@
...
+		users: io(`${HOST}${USERS}`, {path: '/connection'})
```
Comment (Jordan141) [inline src/server/frontEnd/src/utils/sockets/ws-routing.js]: Magic string, put in variable

File: src/server/frontEnd/src/App.js [hunk 3]
```diff
@@ -1,14 +1,18 @@
...
 const {BrowserRouter, Route} = require('react-router-dom')
 const {paths} = require('./routes/routes')
 const LOGIN_URL = paths.HOME
 const REGISTER_URL = paths.REGISTER
 const CHAT_URL = paths.CHAT
 
 class App extends React.Component {
+
+	componentDidMount(){
+		console.log('app props', this.props.store)
```
Comment (Jordan141) [inline src/server/frontEnd/src/App.js]: Remove console.log

File: src/server/frontEnd/src/scenes/chatpage/chatpage.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+		}
+	}
+
+	getSelectedRoom() {
+		return this.props.rooms[this.state.selectedRoom]
+	}
+
+	joinToRoom(roomId) {
+		this.props.joinRoom(roomId)
+		//socket.room.emit(protocols.JOIN, {roomId})
```
Comment (Jordan141) [inline src/server/frontEnd/src/scenes/chatpage/chatpage.js]: remove commented code

File: src/server/frontEnd/src/scenes/chatpage/components/userFinder/userFinder.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+const React = require('react')
+const UsersList = require('../usersList/usersList')
+const UserFinderInput = require('./userFinderInput/userFinderInput')
+class UserFinder extends React.Component {
+	constructor() {
+		super()
+		this.runQuery = this.runQuery.bind(this)
+		this.inviteUser = this.inviteUser.bind(this)
+		this.state = {
+			isQuerrying: false
```
Comment (Jordan141) [inline src/server/frontEnd/src/scenes/chatpage/components/userFinder/userFinder.js]: Typo, should be isQuerying, however this is a non-issue

File: src/server/frontEnd/src/scenes/chatpage/components/userFinder/userFinderInput/userFinderInput.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+	didCeasedToQuery(previousProps) {
+		return previousProps.isQuerrying && !this.props.isQuerrying
+	}
+
+	render() {
+		return (
+			<input
+				className='form-control'
+				palceholder='Username'
```
Comment (Jordan141) [inline src/server/frontEnd/src/scenes/chatpage/components/userFinder/userFinderInput/userFinderInput.js]: typo, should be placeholder

**Comment by Jordan141** [comment]
### More required changes:
- Update the readme.md in this PR to align it with the new way of launching the application. (Both server and frontend)
- Logout button does nothing on click, at least, visually.
- Join room on empty input adds an empty room with no ID

**Comment by michaelKurowski** [comment]
Logout button fixed

**Comment by michaelKurowski** [comment]
Fixed bundle.css issue

**Comment by michaelKurowski** [comment]
User can no more join empty rooms

**Comment by Jordan141** [comment]
### Final changes required:
- [ ] Update the readme.md in this PR to align it with the new way of launching the application. (Both server and frontend).
- [ ] Remove the console.log that's apparently in session.reducer.js:14

**Comment by michaelKurowski** [comment]
TODO: Make E2E tests independent from eachother. Clear Mongo after every test scenario and restart LokIM server

File: src/server/.tests/cypress/integration/chatpage/login.spec.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+			failOnStatusCode: false
+		}).then(() => done())
+	})
+
+	it('can register', () => {
+		cy.get('.home-button.btn.btn-secondary a').click()
+		cy.url().should('include', '/register')
+		cy.get('[name="username"]').type(suite.CORRECT_USERNAME)
+		cy.get('[name="password"]').type(suite.CORRECT_PASSWORD)
+		cy.get('[name="email"]').type('testEmail@test.co.uk')
```
Comment (MarcinGrzeszczak) [inline src/server/.tests/cypress/integration/chatpage/login.spec.js]: move email to suite

File: src/server/.tests/cypress/integration/chatpage/login.spec.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
+/// <reference types="Cypress" />
+let suite
+context('Login', () => {
+	beforeEach(() => {
+		suite = {
+			CORRECT_USERNAME: 'test_username',
+			CORRECT_PASSWORD: 'test_password',
+			WRONG_USERNAME: 'test_user222'
```
Comment (MarcinGrzeszczak) [inline src/server/.tests/cypress/integration/chatpage/login.spec.js]: what happened if user enter wrong password ?

File: src/server/frontEnd/src/.tests/theme/scenes/App.spec.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+		const elementsCount = element.render().length
+		expect(elementsCount).toBe(EXPECTED_ELEMENTS_COUNT)
+	})
+
+	it('lands on homepage by default', () => {
+		const EXPECTED_ID = 'homepage'
+		const element = mount(createApp(suite.store))
+		const elementIdentifier = element.render().children()[0].attribs.id
+		expect(elementIdentifier).toBe(EXPECTED_ID)
+		console.log(suite.store)
```
Comment (MarcinGrzeszczak) [inline src/server/frontEnd/src/.tests/theme/scenes/App.spec.js]: remove console log

File: src/server/frontEnd/src/.tests/theme/scenes/chatpage.spec.js [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+
+	describe('Healthcheck', () => {
+		it('redirects to homepage when user is not logged in', () => {
+			const EXPECTED_ELEMENTS_COUNT = 1
+			const renderedTree = mount(suite.wrapper)
+			const elementsCount = renderedTree.find(suite.HomeMock).length
+			expect(elementsCount).toBe(EXPECTED_ELEMENTS_COUNT)
+		})
+
+		it('redirects to homepage when user is logged in', () => {
```
Comment (MarcinGrzeszczak) [inline src/server/frontEnd/src/.tests/theme/scenes/chatpage.spec.js]: change homepage to chatpage

File: src/server/.tests/cypress/integration/chatpage/login.spec.js (hunk 7 above)
Comment (michaelKurowski) [reply src/server/.tests/cypress/integration/chatpage/login.spec.js]: k

File: src/server/.tests/cypress/integration/chatpage/login.spec.js (hunk 8 above)
Comment (michaelKurowski) [reply src/server/.tests/cypress/integration/chatpage/login.spec.js]: k

File: src/server/frontEnd/src/.tests/theme/scenes/chatpage.spec.js (hunk 10 above)
Comment (michaelKurowski) [reply src/server/frontEnd/src/.tests/theme/scenes/chatpage.spec.js]: ok

File: src/server/frontEnd/src/.tests/theme/scenes/App.spec.js (hunk 9 above)
Comment (michaelKurowski) [reply src/server/frontEnd/src/.tests/theme/scenes/App.spec.js]: ok

**Pull request merged**